contracts/libraries/SafeMigration.sol

```diff
@@ -0,0 +1,136 @@
+import {ISafe} from "../interfaces/ISafe.sol";
+
+/**
+ * @title Migration Contract for Safe Upgrade
+ * @notice This is a generic contract that facilitates the migration of a Safe and SafeL2 contracts.
+ *         The supported target Safe version codehash is immutable and set in the constructor during the deployment of the contract.
+ *         This contract also supports migration with fallback handler update.
+ * @author @safe-global/safe-protocol
+ * @dev IMPORTANT: The library is intended to be used with the Safe standard proxy that stores the singleton address
+ *      at the storage slot 0. Use at your own risk with custom proxy implementations. The contract will allow invocations
+ *      to the migration functions only via delegatecall.
+ */
+contract SafeMigration is SafeStorage {
+    // Address of this contract
+    address public immutable MIGRATION_SINGLETON;
+    // codehash of the Safe singleton implementation
+    bytes32 public immutable SAFE_SINGLETON_CODEHASH;
+    // codehash of the Safe singleton (L2) implementation
+    bytes32 public immutable SAFE_L2_SINGLETON_CODEHASH;
+    // codehash of the fallbackhandler
+    bytes32 public immutable SAFE_FALLBACK_HANDLER_CODEHASH;
+
+    /**
+     * @notice Event indicating a change of master copy address.
+     * @param singleton New master copy address
+     */
+    event ChangedMasterCopy(address singleton);
+
+    /**
+     * @notice Modifier to make a function callable via delegatecall only.
+     * If the function is called via a regular call, it will revert.
+     */
+    modifier onlyDelegateCall() {
+        require(address(this) != MIGRATION_SINGLETON, "Migration should only be called via delegatecall");
+        _;
+    }
+
+    /**
+     * @notice Constructor
+     * @param safeSingleton Address of the Safe singleton implementation
+     * @param safeL2Singleton Address of the SafeL2 singleton implementation
+     */
+    constructor(address safeSingleton, address safeL2Singleton, address fallbackHandler) {
+        MIGRATION_SINGLETON = address(this);
+
+        require(isContract(safeSingleton), "Safe Singleton is not deployed");
+        require(isContract(safeL2Singleton), "Safe Singleton (L2) is not deployed");
+        require(isContract(fallbackHandler), "fallback handler is not deployed");
+
+        SAFE_SINGLETON_CODEHASH = getCodehash(safeSingleton);
+        SAFE_L2_SINGLETON_CODEHASH = getCodehash(safeL2Singleton);
+        SAFE_FALLBACK_HANDLER_CODEHASH = getCodehash(fallbackHandler);
+    }
+
+    /**
+     * @notice Migrate the Safe contract to a new Safe singleton implementation.
+     *         Checks whether the given target singleton address codehash matches the expected Safe/SafeL2 singleton codehash.
+     * @param target Address of the new Safe singleton implementation.
+     */
+    function migrateSingleton(address target) public onlyDelegateCall {
+        require(getCodehash(target) == SAFE_SINGLETON_CODEHASH, "Invalid Safe singleton");
```

Comment:
Just wanted to get your opinion on just having a single function and:

```suggestion
        bytes32 codeHash = getCodeHash(target);
        require(codeHash == SAFE_SINGLETON_CODEHASH || codeHash == SAFE_L2_SINGLETON_CODEHASH, "Invalid Safe singleton");
```

Would bring the total functions from 4 to 2.

Note that I'm 100% OK with how it is, just asking the question to make sure that we are making an explicit decision for this.

Reply:
now that we changed the implementation and no longer have target address as singleton, this suggestion no longer holds.